src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -135,16 +143,46 @@ private void InitializeCore()
                 _oleComponent.ModalStateChanged += OnModalStateChanged;
             }
 
-            UpdateStateMachine();
+            // run it from background and fire and forget
+            StartUpdateStateMachine(pollForStatus:true);
         }
 
-        private void UpdateStateMachine()
+        private void StartUpdateStateMachine(bool pollForStatus)
         {
-            AssertIsForeground();
+            // cancel previous state machine update request
+            _cancellationTokenSource.Cancel();
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            var cancellationToken = _cancellationTokenSource.Token;
+
+            // make sure all state machine change work is serialized so that cancellation
+            // doesn't mess the state up.   
+            _lastTask = _lastTask.ContinueWith(async _ =>
+            {
+                await UpdateStateMachineWorkerAsync(pollForStatus, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken, TaskContinuationOptions.LazyCancellation, TaskScheduler.Default).Unwrap();
+        }
 
-            var currentStatus = IsReSharperEnabled();
+        private async Task UpdateStateMachineWorkerAsync(bool pollForStatus, CancellationToken cancellationToken)
```

Comment:
is pollForStatus still needed?

Reply:
No, I'll remove it.  Thanks for catching that.